---
title: "Mucosal Exposure to non-tuberculous mycobacteria elicits B-cell mediated immunity
  against pulmonary tuberculosis"
author: "Taru Dutt"
date: "2022-10-23"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = FALSE, message = FALSE, warning = FALSE)
```

# About experiment

## Effect of non-tuberculous Mycobacteria on the protective efficacy of BCG 

### Description of the data

Tuberculosis (TB) continues to increase worldwide despite vigorous attempts to control it. Bacillus Calmette-Guerin (BCG) is the only licensed vaccine currently available for protection against TB, however, its efficacy is highly variable between countries. It has been hypothesized that BCGâ€™s variable protection is due, among others, to immunological interference by environmental, non-tuberculous mycobacteria (NTM). However, a definitive mechanism has not been identified so far.  Considering the foregoing, We developed a murine model closely resembling the natural history of human exposure to different mycobacterial species, including: 1) BCG vaccination at an early age; 2) exposure to viable NTMs (Mycobacterium avium subsp. avium) via the oral route and 3) maintaining continuous NTM exposure even after TB infection, as occurs in endemic places.  

The presented data has six groups:

1. Saline control: Negative control + Mycobacterium tuberculosis (Mtb) infection
2. NTM 1X10^5CFU/mL: NTM 1X10^5CFU/mL via drinking water + Mtb infection
3. BCG: BCG vaccination without NTM + Mtb infection
5: BCG + NTM 1X10^5CFU/mL: BCG vaccination + NTM 1X10^5 CFU/mL via drinking water + Mtb infection

## About the technique

We have performed 10x spatial visium transciptomics on formaldehyde-fixed paraffin embedded (FFPE) lung tissues at day 120 post Mtb infection. We have seen lymphoid follicles in lungs of mice vaccinated with BCG, exposed with high concentration of NTM and challenged with Mtb. These lymphoid follicles are correlated with decreased Mtb bacterial burden in the lungs and also with increased B cells and anti-Mtb cell lysate IgA aand IgG antibodies. Therefore, to further evaluate the protection mechanisms, we have performed spatial transcriptomics o lung tissues. 

# Now let's start data analysis

## install packages

```{r}
#remove.packages("Seurat")
#install.packages('remotes')
#remotes::install_version("Seurat", version = "4.0.5")
#BiocManager::install("clusterProfiler")
#install.packages("enrichplot")
#installed.packages("patchwork")
#BiocManager::install("SingleR")
#BiocManager::install("celldex")
#BiocManager::install("ensembldb")
#BiocManager::install('limma')
#BiocManager::install("DESeq2")
#organism = "org.Mm.eg.db"
#BiocManager::install(organism)
#BiocManager::install("pathview")
#install.packages("Seurat")
#install.packages("remotes")
#BiocManager::install("rhdf5")
#install.packages("hdf5r")
#BiocManager::install("NeuCA")
#install.packages('devtools')
#install_github("ggjlab/scMCA")
#install.packages("pheatmap")
#BiocManager::install("scater")
#devtools::install_github("sqjin/CellChat")
#BiocManager::install("ComplexHeatmap")
```

## Load libraries

In this step, we are loading the libraries which are required for analysis and plotting of the data.


```{r}
library(Seurat)
library(devtools)
library(ggplot2)
library(patchwork)
library(dplyr)
library(clusterProfiler)
library(enrichplot)
# we use ggplot2 to add x axis labels (ex: ridgeplot)
library(stringr)
library(celldex)
library(ensembldb)
library(SingleR)
library(DESeq2)
library(org.Mm.eg.db)
library(rhdf5)
library(hdf5r)
library(pheatmap)
library(ComplexHeatmap)
library(CellChat)
library(patchwork)
options(stringsAsFactors = FALSE)
library(NMF)
library(ggalluvial)

```

## load data for mouse 1 and mouse 2

```{r}
# Mouse 1
ntm_bcg_mouse_1 <- "/Volumes/Seagate_4TB/spatial_ntm_samples/demultiplexed/ntm_bcg_sample_round_one/outs/"

list.files(ntm_bcg_mouse_1)

ntm_bcg_mouse_1_data <- Load10X_Spatial(
  ntm_bcg_mouse_1,
  filename = "filtered_feature_bc_matrix.h5",
  assay = "spatial",
  slice = "ntmbcg",
  filter.matrix = TRUE,
  to.upper = FALSE,
  image = NULL)

# Mouse 2
ntm_bcg_mouse_2 <- "/Volumes/Seagate_4TB/spatial_ntm_samples/demultiplexed/ntm_bcg_sample_round_two/outs/"

list.files(ntm_bcg_mouse_2)

ntm_bcg_mouse_2_data <- Load10X_Spatial(
 ntm_bcg_mouse_2,
  filename = "filtered_feature_bc_matrix.h5",
  assay = "spatial",
  slice = "ntmbcg",
  filter.matrix = TRUE,
  to.upper = FALSE,
  image = NULL)


```

## data preprocessing
SCTransform function will NormalizeData, FindVariableFeatures, and ScaleData workflow. Results are saved in a new assay (named SCT by default) with counts being (corrected) counts, and data being log1p(counts).

```{r}
# Mouse 1
ntm_bcg_mouse_1_data <- SCTransform(ntm_bcg_mouse_1_data, assay = "spatial", verbose = FALSE)

# Mouse 2
ntm_bcg_mouse_2_data <- SCTransform(ntm_bcg_mouse_2_data, assay = "spatial", verbose = FALSE)
```

## Visualization of lymphoid follicle markers
### Mouse 1

```{r, fig.width = 10, fig.height= 10}
p1 <- SpatialFeaturePlot(ntm_bcg_mouse_1_data, features = "Cd19", alpha = c(0.1, 1.5), pt.size.factor = 1.8)
p2 <- SpatialFeaturePlot(ntm_bcg_mouse_1_data, features = "Igha", alpha = c(0.1, 1), pt.size.factor = 1.8)
p3 <- SpatialFeaturePlot(ntm_bcg_mouse_1_data, features = "Bcl6", alpha = c(0.1, 1), pt.size.factor = 1.8)
p4 <- SpatialFeaturePlot(ntm_bcg_mouse_1_data, features = "Lta", alpha = c(0.1, 1.5), pt.size.factor = 1.8)
p5 <- SpatialFeaturePlot(ntm_bcg_mouse_1_data, features = "Ltb", alpha = c(0.1, 1.5), pt.size.factor = 1.8)
p6 <- SpatialFeaturePlot(ntm_bcg_mouse_1_data, features = "Cd4", alpha = c(0.1, 1.5), pt.size.factor = 1.8)
p7 <- SpatialFeaturePlot(ntm_bcg_mouse_1_data, features = "Mki67", alpha = c(0.1, 1.5), pt.size.factor = 1.8)
p8 <- SpatialFeaturePlot(ntm_bcg_mouse_1_data, features = "Cxcr5", alpha = c(0.1, 1.5), pt.size.factor = 1.8)

p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8
```

### Mouse 2

```{r, fig.width = 10, fig.height= 10}

p1 <- SpatialFeaturePlot(ntm_bcg_mouse_2_data, features = "Cd19", alpha = c(0.1, 1.5), pt.size.factor = 1.5)
p2 <- SpatialFeaturePlot(ntm_bcg_mouse_2_data, features = "Igha", alpha = c(0.1, 1), pt.size.factor = 1.5)
p3 <- SpatialFeaturePlot(ntm_bcg_mouse_2_data, features = "Bcl6", alpha = c(0.1, 1), pt.size.factor = 1.5)
p4 <- SpatialFeaturePlot(ntm_bcg_mouse_2_data, features = "Lta", alpha = c(0.1, 1.5), pt.size.factor = 1.5)
p5 <- SpatialFeaturePlot(ntm_bcg_mouse_2_data, features = "Ltb", alpha = c(0.1, 1.5), pt.size.factor = 1.5)
p6 <- SpatialFeaturePlot(ntm_bcg_mouse_2_data, features = "Cd4", alpha = c(0.1, 1.5), pt.size.factor = 1.5)
p7 <- SpatialFeaturePlot(ntm_bcg_mouse_2_data, features = "Mki67", alpha = c(0.1, 1.5), pt.size.factor = 1.5)
p8 <- SpatialFeaturePlot(ntm_bcg_mouse_2_data, features = "Cxcr5", alpha = c(0.1, 1.5), pt.size.factor = 1.5)

p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8
```

## dimensionality reduction and clustering

Clustering is the assignment of objects to homogeneous groups (called clusters) while making sure that objects in different groups are not similar. Clustering is considered an unsupervised task as it aims to describe the hidden structure of the objects.

Dimensionality reduction is a process to reduce the number of features under consideration, where each feature is a dimension that partly represents the objects.
Why is dimensionality reduction important? As more features are added, the data becomes very sparse and analysis suffers from the curse of dimensionality. Additionally, it is easier to process smaller data sets.
Dimensionality reduction can be executed using two different methods:

1. Selecting from the existing features (feature selection)
2. Extracting new features by combining the existing features (feature extraction)


```{r, fig.width = 10, fig.height= 10}
# Mouse 1
ntm_bcg_mouse_1_analyzed <- RunPCA(ntm_bcg_mouse_1_data, assay = "SCT", verbose = FALSE)
ntm_bcg_mouse_1_analyzed <- FindNeighbors(ntm_bcg_mouse_1_analyzed, reduction = "pca", dims = 1:5)
ntm_bcg_mouse_1_analyzed <- FindClusters(ntm_bcg_mouse_1_analyzed, verbose = FALSE, resolution = 0.8)
ntm_bcg_mouse_1_analyzed <- RunUMAP(ntm_bcg_mouse_1_analyzed, reduction = "pca", dims = 1:5)

# Mouse 2
ntm_bcg_mouse_2_analyzed <- RunPCA(ntm_bcg_mouse_2_data, assay = "SCT", verbose = FALSE)
ntm_bcg_mouse_2_analyzed <- FindNeighbors(ntm_bcg_mouse_2_analyzed, reduction = "pca", dims = 1:5)
ntm_bcg_mouse_2_analyzed <- FindClusters(ntm_bcg_mouse_2_analyzed, verbose = FALSE, resolution = 0.8)
ntm_bcg_mouse_2_analyzed <- RunUMAP(ntm_bcg_mouse_2_analyzed, reduction = "pca", dims = 1:5)
```


## labelling clusters using SingleR pipeline

```{r}
# First convert the seurat object to single cell experiment object (otherwise the SingleR pipeline will not work)

# Mouse 1
ntm_bcg_mouse_1_analyzed.sce <- as.SingleCellExperiment(ntm_bcg_mouse_1_analyzed)

# Mouse 2
ntm_bcg_mouse_2_analyzed.sce <- as.SingleCellExperiment(ntm_bcg_mouse_2_analyzed)


#create reference data

ref.data <- celldex::ImmGenData()
ref.data

# run singleR pipeline to find the cell types

## Mouse 1
cell_types_mouse_1 <- SingleR(test=ntm_bcg_mouse_1_analyzed.sce, assay.type.test=1, 
    ref=ref.data, labels=ref.data$label.main)

## Mouse 2
cell_types_mouse_2 <- SingleR(test=ntm_bcg_mouse_2_analyzed.sce, assay.type.test=1, 
    ref=ref.data, labels=ref.data$label.main)

```

## merging metadata of cell types to seurat object

```{r}
# Mouse 1
ntm_bcg_mouse_1_analyzed[["ref.data"]] <- cell_types_mouse_1$labels

# Mouse 2
ntm_bcg_mouse_2_analyzed[["ref.data"]] <- cell_types_mouse_2$labels
```

## Plot umap and spatial plot to localize each cell type

### Mouse 1

```{r, fig.width = 16, fig.height= 10}


remove.packages("Seurat")
remotes::install_version(package = 'Seurat', version = package_version('4.0.0'))
library(Seurat)


p1 <- DimPlot(ntm_bcg_mouse_1_analyzed, group.by = c("ref.data"), reduction = "umap",label = F, pt.size = 1.5, label.size = 5, cols = my_cols)

p2 <- SpatialDimPlot(ntm_bcg_mouse_1_analyzed, group.by = "ref.data", label = FALSE, label.size = 5, cols = my_cols)

p1 + p2
```


### Mouse 2

```{r, fig.width = 16, fig.height= 10}

p1 <- DimPlot(ntm_bcg_mouse_2_analyzed, group.by = c("ref.data"), reduction = "umap",label = F, pt.size = 1.5, label.size = 5, cols = my_cols)

p2 <- SpatialDimPlot(ntm_bcg_mouse_2_analyzed, group.by = c("ref.data"), label = FALSE, label.size = 5, cols = my_cols)

p1 + p2
```

## Plot violin plot with top markers associated with B cells and lymphoid follicles

### Mouse 1

```{r, fig.width = 14, fig.height= 12}
VlnPlot(ntm_bcg_mouse_1_analyzed, group.by = c("ref.data"), features = c("Cxcr5", "Mki67", "Ltb", "Cd38", "Ccr6", "Tnfrsf13c", "Fcrla"), cols =  my_cols)

```

### Mouse 2

```{r, fig.width = 14, fig.height= 12}
VlnPlot(ntm_bcg_mouse_2_analyzed, group.by = c("ref.data"), features = c("Cxcr5", "Mki67", "Ltb", "Cd38", "Ccr6", "Tnfrsf13c", "Fcrla"), cols =  my_cols)
```

## find markers in each cluster

### Mouse 1
```{r}

B_cell_mouse_1 <- FindMarkers(ntm_mouse_1_analyzed, group.by = "ref.data", ident.1 = "B cells", min.pct = 0.25)
head(B_cell_mouse_1, n = 10)
write.csv(B_cell_mouse_1, "DATA/bcg_ntm/B_cell_mouse_1.csv")

DC_mouse_1 <- FindMarkers(ntm_mouse_1_analyzed, group.by = "ref.data",  ident.1 = "DC", min.pct = 0.25)
head(DC_mouse_1, n = 10)
write.csv(DC_mouse_1, "DATA/bcg_ntm/DC_mouse_1.csv")

Stromal_cells_mouse_1 <- FindMarkers(ntm_mouse_1_analyzed, group.by = "ref.data",  ident.1 = "Stromal cells", min.pct = 0.25)
head(Stromal_cells_mouse_1, n = 10)
write.csv(Stromal_cells_mouse_1, "DATA/bcg_ntm/Stromal_cells_mouse_1.csv")


Endothelial_cells_mouse_1 <- FindMarkers(ntm_mouse_1_analyzed, group.by = "ref.data",  ident.1 = "Endothelial cells", min.pct = 0.25)
head(Endothelial_cells_mouse_1, n = 10)
write.csv(Endothelial_cells_mouse_1, "DATA/bcg_ntm/Endothelial_cells_mouse_1.csv")

Epithelial_cell_mouse_1 <- FindMarkers(ntm_mouse_1_analyzed, group.by = "ref.data",  ident.1 = "Epithelial cells", min.pct = 0.25)
head(Epithelial_cell_mouse_1, n = 10)
write.csv(Epithelial_cell_mouse_1, "DATA/bcg_ntm/Epithelial_cell_mouse_1.csv")

Fibroblasts_mouse_1 <- FindMarkers(ntm_mouse_1_analyzed, group.by = "ref.data",  ident.1 = "Fibroblasts", min.pct = 0.25)
head(Fibroblasts_mouse_1, n = 10)
  write.csv(Fibroblasts_mouse_1, "DATA/bcg_ntm/Fibroblasts_mouse_1.csv")
  
Macrophages_mouse_1 <- FindMarkers(ntm_mouse_1_analyzed, group.by = "ref.data",  ident.1 = "Macrophages", min.pct = 0.25)
  head(Macrophages_mouse_1, n = 10)
write.csv(Macrophages_mouse_1, "DATA/bcg_ntm/Macrophages_mouse_1.csv")

Monocytes_mouse_1 <- FindMarkers(ntm_mouse_1_analyzed, group.by = "ref.data",  ident.1 = "Monocytes", min.pct = 0.25)
  head(Monocytes_mouse_1, n = 10)
write.csv(Monocytes_mouse_1, "DATA/bcg_ntm/Monocytes_mouse_1.csv")


```

### Mouse 2
```{r}

B_cell_mouse_2 <- FindMarkers(ntm_mouse_2_analyzed, group.by = "ref.data", ident.1 = "B cells", min.pct = 0.25)
head(B_cell_mouse_2, n = 10)
write.csv(B_cell_mouse_2, "DATA/bcg_ntm/B_cell_mouse_2.csv")

DC_mouse_2 <- FindMarkers(ntm_mouse_2_analyzed, group.by = "ref.data",  ident.1 = "DC", min.pct = 0.25)
head(DC_mouse_2, n = 10)
write.csv(DC_mouse_2, "DATA/bcg_ntm/DC_mouse_2.csv")

Stromal_cells_mouse_2 <- FindMarkers(ntm_mouse_2_analyzed, group.by = "ref.data",  ident.1 = "Stromal cells", min.pct = 0.25)
head(Stromal_cells_mouse_2, n = 10)
write.csv(Stromal_cells_mouse_2, "DATA/bcg_ntm/Stromal_cells_mouse_2.csv")


Endothelial_cells_mouse_2 <- FindMarkers(ntm_mouse_2_analyzed, group.by = "ref.data",  ident.1 = "Endothelial cells", min.pct = 0.25)
head(Endothelial_cells_mouse_2, n = 10)
write.csv(Endothelial_cells_mouse_2, "DATA/bcg_ntm/Endothelial_cells_mouse_2.csv")

Epithelial_cell_mouse_2 <- FindMarkers(ntm_mouse_2_analyzed, group.by = "ref.data",  ident.1 = "Epithelial cells", min.pct = 0.25)
head(Epithelial_cell_mouse_2, n = 10)
write.csv(Epithelial_cell_mouse_2, "DATA/bcg_ntm/Epithelial_cell_mouse_2.csv")

Fibroblasts_mouse_2 <- FindMarkers(ntm_mouse_2_analyzed, group.by = "ref.data",  ident.1 = "Fibroblasts", min.pct = 0.25)
head(Fibroblasts_mouse_2, n = 10)
  write.csv(Fibroblasts_mouse_2, "DATA/bcg_ntm/Fibroblasts_mouse_2.csv")
  
Macrophages_mouse_2 <- FindMarkers(ntm_mouse_2_analyzed, group.by = "ref.data",  ident.1 = "Macrophages", min.pct = 0.25)
  head(Macrophages_mouse_2, n = 10)
write.csv(Macrophages_mouse_2, "DATA/bcg_ntm/Macrophages_mouse_2.csv")

Monocytes_mouse_2 <- FindMarkers(ntm_mouse_2_analyzed, group.by = "ref.data",  ident.1 = "Monocytes", min.pct = 0.25)
  head(Monocytes_mouse_2, n = 10)
write.csv(Monocytes_mouse_2, "DATA/bcg_ntm/Monocytes_mouse_2.csv")

Neutrophils_mouse_2 <- FindMarkers(ntm_mouse_2_analyzed, group.by = "ref.data",  ident.1 = "Neutrophils", min.pct = 0.25)
  head(Neutrophils_mouse_2, n = 10)
write.csv(Neutrophils_mouse_2, "DATA/bcg_ntm/Neutrophils_mouse_2.csv")
```

```{r}
B_cell_deseq_1 <- FindMarkers(ntm_bcg_mouse_1_analyzed, group.by = "ref.data",  ident.1 = "B cells", test.use = "DESeq2")
write.csv(B_cell_deseq_1, "DATA/bcg_ntm/B_cell_deseq_mouse1.csv")

B_cell_deseq_2 <- FindMarkers(ntm_bcg_mouse_2_analyzed, group.by = "ref.data",  ident.1 = "B cells", test.use = "DESeq2")
write.csv(B_cell_deseq_2, "DATA/bcg_ntm/B_cell_deseq_mouse2.csv")
```

# Gene set enrichment

```{r}
organism = org.Mm.eg.db
```

## prepare input data for gene set enrichment

```{r}
# reading in data from deseq2
df = read.csv("DATA/bcg_ntm/B_cell_deseq.csv", header=TRUE)

# we want the log2 fold change 
original_gene_list <- df$avg_log2FC

# name the vector
names(original_gene_list) <- df$X

# omit any NA values 
gene_list<-na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)
```

## Gene set enrichment

```{r}
#gene set enrichment function for B cell cluster

gse <- gseGO(geneList=gene_list, 
             ont ="ALL", 
             keyType = "SYMBOL", 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = organism, 
             pAdjustMethod = "none",
             eps = 0.0,
             nPermSimple = 10000)

dotplot(gse, showCategory=10, split=".sign") + facet_grid(.~.sign)

```

## KEGG pathway analysis

```{r}
# Convert gene IDs for gse to KEGG function
# We will lose some genes here because not all IDs will be converted

ids<-bitr(names(gene_list), fromType = "SYMBOL", toType = "ENTREZID", OrgDb=organism)
 
# remove duplicate IDS (here I use "ENSEMBL", but it should be whatever was selected as keyType)
dedup_ids = ids[!duplicated(ids[c("SYMBOL")]),]

# Create a new dataframe df2 which has only the genes which were successfully mapped using the bitr function above
df2 = df[df$X %in% dedup_ids$SYMBOL,]

# Create a new column in df2 with the corresponding ENTREZ IDs
df2$Y = dedup_ids$ENTREZID

# Create a vector of the gene unuiverse
kegg_gene_list <- df2$avg_log2FC

# Name vector with ENTREZ ids
names(kegg_gene_list) <- df2$Y

# omit any NA values 
kegg_gene_list<-na.omit(kegg_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
kegg_gene_list = sort(kegg_gene_list, decreasing = TRUE)
```

## running kegg

```{r}
kegg_organism = "mmu"
kk2 <- gseKEGG(geneList     = kegg_gene_list,
               organism     = kegg_organism,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid",
               eps = 0.0)

write.csv(kk2, "DATA/bcg_ntm/kk2.csv")

dotplot(kk2, showCategory=, split=".sign") + facet_grid(.~.sign)
```


```{r}
library(pathview)

# Produce the native KEGG plot (PNG) for first set of genes
dme <- pathview(gene.data=kegg_gene_list, pathway.id="mmu05417", species = kegg_organism, low = list(gene = "#453781FF"), mid =
list(gene = "#20A387FF"), high = list(gene = "#FDE725FF"))

# Produce a different plot (PDF) (not displayed here)
dme <- pathview(gene.data=kegg_gene_list, pathway.id="mmu05417", species = kegg_organism, kegg.native = F, low = list(gene = "#453781FF"), mid =
list(gene = "#20A387FF"), high = list(gene = "#FDE725FF"))

```
